cpp/ql/src/semmle/code/cpp/Namespace.qll

```diff
@@ -150,7 +153,7 @@ class UsingDeclarationEntry extends UsingEntry {
    */
   Declaration getDeclaration() { usings(underlyingElement(this), unresolveElement(result), _) }
 
-  override string toString() { result = "using " + this.getDeclaration().toString() }
+  override string toString() { result = "using declaration" }
```

Comment:
Could we use `this.getDeclaration().getQualifiedName()` here? I think the performance problems with `getQualifiedName` were only when it was being used for string comparison, not with the `Declaration` bound.

Reply:
This is a good idea, I'll look into it next week.

Reply:
I don't understand @rdmarsh2's comment. The `Declaration.getQualifiedName` predicate is very expensive to compute, and if we can avoid computing it in this stage then we'll have better performance.

To make this non-recursive without changing its output, you can move the body of `Declaration.toString` into a new predicate `Declaration.getDescription` and then call `Declaration.getDescription` both from here and from `Declaration.toString`.

Reply:
Yes, I haven't seen the big performance cost but it turns out `getQualifiedName` doesn't match the `toString` anyway (in this case).  Since I'm trying to minimize the number of result changes this causes I'm going with @jbj's solution.